Document constructor auth for register/register_at (#2666)
stellar/stellar-docs: docs/build/guides/testing/test-contract-auth.mdx

For the full example the above snippet is extracted from, see the [auth example contract](../../smart-contracts/example-contracts/auth.mdx).

:::

## Authorization in constructors

> If a contract's constructor calls `require_auth()` (or `require_auth_for_args()`), both `env.register(...)` and `env.register_at(...)` switch the environment to recording-auth mode, inherited from the current auth manager, for the duration of the constructor invocation, then restore the previous auth manager afterward. This means constructor auth checks succeed for contracts registered with either function, without requiring `mock_all_auths()` to be called first.

Reviewer: the first sentence here feels a little long and complex. something like this might offer a bit more clarity:

```suggestion
If a contract's constructor calls `require_auth()` (or `require_auth_for_args()`), both `env.register(...)` and `env.register_at(...)` switch the environment to recording-auth mode for the duration of the constructor invocation, inheriting the mode from the current auth manager. Once the constructor returns, the previous auth manager is restored. This means constructor auth checks succeed for contracts registered with either function, without requiring `mock_all_auths()` to be called first.
```

To have auth in a constructor execute as it will in production the contract must be deployed in the test using the standard deployment functions after being built to Wasm.

:::caution

> This behaviour was not consistent in some versions of the `soroban-sdk` due to an issue, but is consistent as of v27.0.2.

Reviewer: since we're pointing out an issue that exists, should we make `an issue` a link to that github issue?

i'm not insisting on it, just asking if it might add some extra context for the curious.

Author: The issue has been resolved so I don't think it is useful to link to it.
